ares_gethostbyname.c

```diff
@@ -118,11 +134,95 @@ void ares_gethostbyname(ares_channel channel, const char *name, int family,
   hquery->arg = arg;
   hquery->remaining_lookups = channel->lookups;
   hquery->timeouts = 0;
+  hquery->next_domain = 0;
+  hquery->ever_got_data = 0;
 
   /* Start performing lookups according to channel->lookups. */
   next_lookup(hquery, ARES_ECONNREFUSED /* initial error code */);
 }
 
+static void next_dns_lookup(struct host_query *hquery, int status) {
+  char *s = NULL;
+  int is_s_allocated = 0;
+  int next_d = 0;
+
+  /* What is the next family we try? */
+  if (hquery->sent_family == -1)
+    {
+      if (hquery->want_family == AF_UNSPEC || hquery->want_family == AF_INET6)
+          hquery->sent_family = AF_INET6;
+      else
+          hquery->sent_family = AF_INET;
+    }
+  else
+    {
+      if (hquery->want_family == AF_UNSPEC)
+        {
+          next_d = hquery->sent_family == AF_INET;
+          hquery->sent_family = hquery->sent_family == AF_INET6 ? AF_INET : AF_INET6;
+        }
+      else
+        {
+          next_d = 1;
+          hquery->sent_family = hquery->want_family;
+        }
+    }
+
+  if (next_d && hquery->ever_got_data)
+    {
+      /* No more queries on another domain to check since
+         we already made a successful query for another type
+      */
+      free_hquery(hquery, NULL);
```

Comment:
Thanks for pointing this out. `free_hquery` should only be called if a previous `hquery->callback` was made (T_AAAA) and the second family request (T_A) yield an unsuccessful response. Then the query could silently be closed because callback was already called for T_AAAA. I will correct this. `ever_got_data` should store this info which is currently not correct implemented. I will work on it.